Restricted kube object name sizes.

mook-as opened the pull request:
Kubernetes doesn't like resource names longer than 63 characters.  Try to make it so all the names are shorter than that (by including part of the hash of the full thing).

Co-authored-by: Mark Yen <mark.yen@suse.com>

Commit:
Restricted kube object name sizes.

Co-authored-by: Mark Yen <mark.yen@suse.com>

jandubois commented on kube/bosh_deployment_manifest_secret.go:
```diff
@@ -9,7 +13,15 @@ func MakeBoshDeploymentManifestSecret(settings ExportSettings) (helm.Node, error
 	}
 
-	secret := newKubeConfig(settings, "v1", "Secret", "deployment-manifest")
+	cb := NewConfigBuilder().
```
In general I would prefer to have refactoring like this in a separate commit from the new/changed functionality. Don't bother "fixing" it for this PR; just a request for future changes.

jandubois commented on kube/bosh_deployment_manifest_secret.go:
```diff
@@ -9,7 +13,15 @@ func MakeBoshDeploymentManifestSecret(settings ExportSettings) (helm.Node, error
 
-	secret := newKubeConfig(settings, "v1", "Secret", "deployment-manifest")
+	cb := NewConfigBuilder().
+		SetSettings(&settings).
```
Why are `settings` passed by reference and not by value, like everywhere else? I would not expect `SetSettings` to change its argument.

f0rmiga replied:
Passing the reference doesn't mean it's going to be changed.

jandubois commented on kube/kube_test.go:
```diff
@@ -88,12 +94,24 @@ func RenderNode(node helm.Node, config interface{}) ([]byte, error) {
 	// Note: Replicate helm's behaviour on missing keys.
-	tmpl, err := template.New("").Option("missingkey=zero").Funcs(functions).Parse(string(helmConfig.Bytes()))
+	tmpl := template.New("").Option("missingkey=zero").Funcs(functions)
+	if err != nil {
```
Where is `err` being set here?

mook-as replied:
It's not; my mistake. Sorry about that.

jandubois commented on kube/rbac.go:
```diff
@@ -48,12 +48,19 @@ func NewRBACAccount(accountName string, config *model.Configuration, settings Ex
+			SetName(accountName).
+			AddModifier(block).
+			AddModifier(helm.Comment(description))
+		resource, err := cb.Build()
```
Why not call it `serviceAccount` instead of `resource`? That leads to the more readable

```
resources = append(resources, serviceAccount)
```

instead of the eye-squinting

```
resources = append(resources, resource)
```

jandubois commented on kube/rbac.go:
```diff
@@ -122,23 +134,39 @@ func NewRBACAccount(accountName string, config *model.Configuration, settings Ex
+			SetSettings(&settings).
+			SetAPIVersion("rbac.authorization.k8s.io/v1").
+			SetKind("ClusterRoleBinding").
+			SetName(fmt.Sprintf("%s-%s-cluster-binding", accountName, clusterRoleName)).
```
Don't care too much, but maybe set the `Name` in the `else` branch of `if settings.CreateHelmChart` below, for symmetry with how you set the `roleRef` name a little further down?

jandubois commented on kube/templates.go:
```diff
@@ -0,0 +1,29 @@
+		{{ end }}`
+	return []helm.Node{
+		helm.NewNode(
+			regexp.MustCompile(`(^|\}\})\s+(\{\{|$)`).
```
Is this really necessary? I just spent 3 minutes figuring out what the regexp is supposed to do and to convince myself that is correct and not doing too much or too little. At a minimum it should have a comment on top.

It would be much simpler/obvious to use the templating mechanism for eating unwanted whitespace: `{{- expre -}}`.

jandubois commented on kube/utils.go:
```diff
@@ -39,9 +39,84 @@ func newSelector(role *model.InstanceGroup, settings ExportSettings) *helm.Mappi
-// newKubeConfig sets up generic a Kube config structure with minimal metadata
-func newKubeConfig(settings ExportSettings, apiVersion, kind, name string, modifiers ...helm.NodeModifier) *helm.Mapping {
-	labels := helm.NewMapping(RoleNameLabel, name) // "app.kubernetes.io/component"
+// ConfigBuilder sets up generic a Kube resource structure with minimal metadata.
```
```suggestion
// ConfigBuilder sets up a generic Kube resource structure with minimal metadata.
```

jandubois commented on kube/utils.go:
```diff
@@ -39,9 +39,84 @@ func newSelector(role *model.InstanceGroup, settings ExportSettings) *helm.Mappi
+	if b.err == nil {
+		b.err = err
+	} else {
+		b.err = fmt.Errorf("%v: %v", b.err, err)
```
Why `:` instead of `,`; the errors can be independent of each other. The `:` implies that the right side provides more details about the left.

Nit-pickers-r-us

f0rmiga replied:
Thanks for pointing this out. You are absolutely right, my bad.

jandubois commented on kube/utils.go:
```diff
@@ -39,9 +39,84 @@ func newSelector(role *model.InstanceGroup, settings ExportSettings) *helm.Mappi
+	} else if !b.settings.CreateHelmChart {
+		b.setError(fmt.Errorf("name is a Helm expression, but not creating a helm chart"))
+	}
+	b.name = name
```
You could add a check here that `name` starts with `{{`. Otherwise `helm/config.go` might put quotes around the expression. Not sure if it matters in this specific case.

jandubois reviewed (commented):
None of the comments totally require changes, but at least the unset `err` thing should be fixed, I think. (It is going to work as-is, because `err` is always `nil` at that point, but then the whole condition doesn't make sense).

mook-as committed:
Address comments

https://github.com/cloudfoundry-incubator/fissile/pull/465#pullrequestreview-195167375

Note that GetHelmTemplateHelpers() can't use a literal string with new
lines because `helm.Scalar` would escape them on output as `\n` (as in,
a two-character string), which breaks the go template whitespace-eating.
